src/Features/Core/Portable/DesignerAttributes/AbstractDesignerAttributeService.cs

```diff
@@ -140,5 +101,44 @@ private static string GetArgumentString(TypedConstant argument)
 
             return ((string)argument.Value).Trim();
         }
+
+        internal static async Task<ImmutableDictionary<string, DesignerAttributeDocumentData>> TryAnalyzeProjectInCurrentProcessAsync(
+            Project project, CancellationToken cancellationToken)
+        {
+            var projectVersion = await project.GetDependentVersionAsync(cancellationToken).ConfigureAwait(false);
+            var semanticVersion = await project.GetDependentSemanticVersionAsync(cancellationToken).ConfigureAwait(false);
+
+            // Get whatever data we've current persisted.
+            var designerAttributeData = await DesignerAttributeProjectData.ReadAsync(
+                project, cancellationToken).ConfigureAwait(false);
+
+            // If we have no persisted data, or the persisted data is for a previous version of 
+            // the project, then compute the results for the current project snapshot.
+            if (designerAttributeData == null ||
+                !VersionStamp.CanReusePersistedVersion(semanticVersion, designerAttributeData.SemanticVersion))
+            {
+                designerAttributeData = await ComputeAndPersistDesignerAttributeProjectDataAsync(
+                    project, semanticVersion, cancellationToken).ConfigureAwait(false);
+            }
+
+            return designerAttributeData.PathToDocumentData;
+        }
+
+        private static async Task<DesignerAttributeProjectData> ComputeAndPersistDesignerAttributeProjectDataAsync(
+            Project project, VersionStamp semanticVersion, CancellationToken cancellationToken)
+        {
+            var service = project.LanguageServices.GetService<IDesignerAttributeService>();
+
+            var builder = ImmutableDictionary.CreateBuilder<string, DesignerAttributeDocumentData>();
```

Comment:
Okay - I actually hadn't realized it was `ImmutableDictionary` and not `ImmutableArray`.  No big deal.  🤷‍♂️